extensions/visualizations/oppia-visualization-click-hexbins.directive.spec.ts

```diff
@@ -0,0 +1,126 @@
+// Copyright 2020 The Oppia Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Directive unit tests for the "click hexbins" visualization.
+ */
+
+// TODO(#7222): Remove the following block of unnecessary imports once
+// the code corresponding to the spec is upgraded to Angular 8.
+import { UpgradedServices } from 'services/UpgradedServices';
+
+require('visualizations/oppia-visualization-click-hexbins.directive.ts');
+
+describe('Oppia click hexbins visualization', function() {
+  let $compile, $rootScope, AssetsBackendApiService, ContextService,
+    ImagePreloaderService;
+  let el: JQLite;
+
+  beforeEach(angular.mock.module('oppia', function($provide) {
+    const ugs = new UpgradedServices();
+    for (let [key, value] of Object.entries(ugs.getUpgradedServices())) {
+      $provide.value(key, value);
+    }
+  }));
```

Comment:
@nithusha21 should Brian follow the Srijan suggestion of not using upgraded services like this? /cc @srijanreddy98

Reply:
I think the solution is to use TestBed.provide instead of ugs. @srijanreddy98 can help you with how to use it. other than that this test file looks great!

Reply:
Hmm it doesn't seem possible since AssetsBackendApiService and ImagePreloaderService are both AngularJS services, so there's no type I can provide to TestBed.

Looks like we'll need to wait until those two services are migrated to Angular 8 first?

Reply:
@brianrodri, I think we can go ahead and remove these lines. @brianrodri can just run the tests without these five lines. I think as you are spying on them, we may not have to provide it.

Reply:
I was able to bypass the errors by using $provide to create mocks. Does this seem OK?